### PR TITLE
Fix sorting timeline by due date

### LINKED ISSUE #764: Program Chair Console: timeline should be sorted chronologically
Screen Shot 2020-10-09 at 10 53 48 AM

### PR DESCRIPTION
**Pull request by haroldrubio**
Closes #764

### CODE REVIEW

File: openreview/conference/templates/programchairWebfield.js [hunk 1]
```diff
@@ -959,31 +959,53 @@ var displayStatsAndConfiguration = function(conferenceStats) {
...
-  html += renderInvitation(invitationMap, CONFERENCE_ID + '/-/' + DECISION_NAME, 'Decisions')
+  push_to_dated_arrays(invitationMap, CONFERENCE_ID + '/-/' + OFFICIAL_REVIEW_NAME, 'Reviewing', dated_elements, not_dated_elements)
+  push_to_dated_arrays(invitationMap, CONFERENCE_ID + '/-/' + COMMENT_NAME, 'Commenting', dated_elements, not_dated_elements)
+  push_to_dated_arrays(invitationMap, CONFERENCE_ID + '/-/' + OFFICIAL_META_REVIEW_NAME, 'Meta Reviews', dated_elements, not_dated_elements)
+  push_to_dated_arrays(invitationMap, CONFERENCE_ID + '/-/' + DECISION_NAME, 'Decisions', dated_elements, not_dated_elements)
+
+  // Sort + append in dated order, followed by not-dated content
+  dated_elements.sort(function (a, b) {return a['date'] - b['date']})
+
+  for (var i = 0; i < dated_elements.length; i++)
```
Comment (zbialecki) [inline openreview/conference/templates/programchairWebfield.js:1004]: Generally, we prefer map and filter for this type of array manipulation. This could be written as:

```js
html += datedElements.map(function(elem) { return elem.content; }).join('\n');
html += undatedElements.join('\n');
```

File: openreview/conference/templates/programchairWebfield.js [hunk 2]
```diff
@@ -959,31 +959,53 @@ var displayStatsAndConfiguration = function(conferenceStats) {
...
-
   html += '<div class="col-md-8 col-xs-12">'
   html += '<h4>Timeline:</h4><ul style="padding-left: 15px">';
-  html += renderInvitation(invitationMap, SUBMISSION_ID, 'Paper Submissions')
-  html += renderInvitation(invitationMap, REVIEWERS_ID + '/-/' + BID_NAME, 'Reviewers Bidding')
+  var not_dated_elements = []
+  var dated_elements = []
+  var push_to_dated_arrays = function(invitationMap, id, name, dated, not_dated){
+    var curr_content = renderInvitation(invitationMap, id, name)
+    if (curr_content)
```
Comment (zbialecki) [inline openreview/conference/templates/programchairWebfield.js:968]: It's better to return early: `if (!currContent) return;`. It keeps things easier to read and prevents conditions from nesting too deeply.

**Review by zbialecki — COMMENTED** [review]
Found the python developer 😉  

JavaScript has slightly different conventions, we generally follow the AirBnB style guide: https://github.com/airbnb/javascript/tree/es5-deprecated/es5 (for webfields all JS has to be valid ES5 code).

That means variables should be camelCase, lines should end in `;` and `if`s should use `{ }` unless you are returning on the same line.

File: openreview/conference/templates/programchairWebfield.js (hunk 1 above)
Comment (haroldrubio) [reply openreview/conference/templates/programchairWebfield.js]: Alright, I've replaced the two for loops with the `map()` and `join()` functions. I wasn't aware that JavaScript had map and reduce functions for generic arrays, that's cool

File: openreview/conference/templates/programchairWebfield.js (hunk 2 above)
Comment (haroldrubio) [reply openreview/conference/templates/programchairWebfield.js]: Yup, I've made those changes

**Comment by haroldrubio** [comment]
Thanks for the feedback! It's been a while since I've used these conventions so I've gotta get back into the habit - I have noticed that some lines do have semicolons while some don't so I'll have to look more into JS convention.

File: openreview/conference/templates/programchairWebfield.js [hunk 3]
```diff
@@ -959,31 +959,54 @@ var displayStatsAndConfiguration = function(conferenceStats) {
...
   // Timeline
   var invitationMap = conferenceStatusData.invitationMap;
-
   html += '<div class="col-md-8 col-xs-12">'
   html += '<h4>Timeline:</h4><ul style="padding-left: 15px">';
-  html += renderInvitation(invitationMap, SUBMISSION_ID, 'Paper Submissions')
-  html += renderInvitation(invitationMap, REVIEWERS_ID + '/-/' + BID_NAME, 'Reviewers Bidding')
+  var notDatedElements = [];
+  var datedElements = [];
+  var pushToDatedArrays = function(invitationMap, id, name, dated, notDated){
```
Comment (zbialecki) [inline openreview/conference/templates/programchairWebfield.js:966]: You don't need to pass references dated and notDated to this function, you can just push to the arrays defined in the outer scope.

File: openreview/conference/templates/programchairWebfield.js [hunk 4]
```diff
@@ -959,31 +959,54 @@ var displayStatsAndConfiguration = function(conferenceStats) {
...
+    var currContent = renderInvitation(invitationMap, id, name);
+    if (!currContent) return;
+
+    var currDate = invitationMap[id].duedate;
+    if (currDate){
```
Comment (zbialecki) [inline openreview/conference/templates/programchairWebfield.js:971]: Curly braces should have spaces around them and else should be on the same line as the closing brace.

File: openreview/conference/templates/programchairWebfield.js [hunk 5]
```diff
@@ -959,31 +959,54 @@ var displayStatsAndConfiguration = function(conferenceStats) {
...
-  html += renderInvitation(invitationMap, CONFERENCE_ID + '/-/' + COMMENT_NAME, 'Commenting')
-  html += renderInvitation(invitationMap, CONFERENCE_ID + '/-/' + OFFICIAL_META_REVIEW_NAME, 'Meta Reviews')
-  html += renderInvitation(invitationMap, CONFERENCE_ID + '/-/' + DECISION_NAME, 'Decisions')
+  pushToDatedArrays(invitationMap, CONFERENCE_ID + '/-/' + OFFICIAL_REVIEW_NAME, 'Reviewing', datedElements, notDatedElements);
+  pushToDatedArrays(invitationMap, CONFERENCE_ID + '/-/' + COMMENT_NAME, 'Commenting', datedElements, notDatedElements);
+  pushToDatedArrays(invitationMap, CONFERENCE_ID + '/-/' + OFFICIAL_META_REVIEW_NAME, 'Meta Reviews', datedElements, notDatedElements);
+  pushToDatedArrays(invitationMap, CONFERENCE_ID + '/-/' + DECISION_NAME, 'Decisions', datedElements, notDatedElements);
+
+  // Sort + append in dated order, followed by not-dated content
+  datedElements.sort(function (a, b) {return a['date'] - b['date']});
```
Comment (zbialecki) [inline openreview/conference/templates/programchairWebfield.js:1005]: In javascript you should use dot notation whenever possible,
```suggestion
  datedElements.sort(function (a, b) { return a.date - b.date; });
```

Comment (haroldrubio) [reply openreview/conference/templates/programchairWebfield.js]: Ah right, that makes sense - I did read that `{ 'content': currContent, 'date': new Date(currDate) }` makes a Javascript object so it makes sense to access them with dots instead of using a "key"

File: openreview/conference/templates/programchairWebfield.js [hunk 6]
```diff
@@ -959,31 +959,54 @@ var displayStatsAndConfiguration = function(conferenceStats) {
...
+    var currContent = renderInvitation(invitationMap, id, name);
+    if (!currContent) return;
+
+    var currDate = invitationMap[id].duedate;
+    if (currDate) {
+      datedElements.push({ 'content': currContent, 'date': new Date(currDate) });
+    }
+    else {
+      notDatedElements.push(currContent);
+    }
```
Comment (zbialecki) [inline openreview/conference/templates/programchairWebfield.js:976]: Just 1 more suggestion:

```suggestion
    if (currDate) {
      datedElements.push({ content: currContent, date: new Date(currDate) });
    } else {
      notDatedElements.push(currContent);
    }
```

**Review by zbialecki — APPROVED** [review]
Looks good!

**Comment by zbialecki** [comment]
@celestemartinez can you review and merge this when you get a chance?

**Review by celestemartinez — APPROVED** [review]

File: tests/test_eccv_conference.py [hunk 7]
```diff
@@ -1655,3 +1656,41 @@ def test_secondary_assignments(self, conference, client, test_client, selenium,
         secondary_assignments_div = selenium.find_element_by_id('secondary-papers')
         for i in range(1,3):
             assert secondary_assignments_div.find_elements_by_id('note-summary-{}'.format(i))
+
+    def test_chronological_pc_timeline(self, conference, selenium, request_page):
+        # Assumptions: 1) Items are in list elements
+        #              2) Dates are stated in plain English
+        def convert_text_to_datetime(text):
+            mnt_to_idx = {month.lower(): index for index, month in enumerate(calendar.month_abbr) if month}
```
Comment (xkopenreview) [inline tests/test_eccv_conference.py:1664]: calendar.month_abbr has "Sep" for september while the one displayed in ui is "Sept"
so ln 1673 would fail
image
